varname/core.py

```diff
@@ -0,0 +1,281 @@
+"""Provide core features for varname"""
+
+import ast
+import warnings
+from typing import Optional, Tuple, Union
+from .utils import (
+    bytecode_nameof,
+    get_node,
+    lookfor_parent_assign,
+    node_name,
+    VarnameRetrievingError,
+    MultiTargetAssignmentWarning
+)
+from .ignore import IgnoreList, IgnoreType
+
+def varname(
+        frame: int = 1,
+        ignore: Optional[IgnoreType] = None,
+        multi_vars: bool = False,
+        raise_exc: bool = True
+) -> Optional[Union[str, Tuple[Union[str, tuple]]]]:
+    """Get the name of the variable(s) that assigned by function call or
+    class instantiation.
+
+    To debug and specify the right frame and ignore arguments, you can set
+    debug on and see how the frames are ignored or selected:
+
+    >>> from varname import config
+    >>> config.debug = True
+
+    Args:
+        frame: `N`th frame used to retrieve the variable name. This means
+            `N-1` intermediate frames will be skipped. Note that the frames
+            match `ignore` will not be counted. See `ignore` for details.
+        ignore: Frames to be ignored in order to reach the `N`th frame.
+            These frames will not be counted to skip within that `N-1` frames.
+            You can specify:
+            - A module (or filename of a module). Any calls from it and its
+                submodules will be ignored.
+            - A function. Directly ignore a function. If it is a suspicious
+                decorated function, a warning will be shown. See decorated
+                function below.
+            - Tuple of a function and number of decorators. The decorators
+                on the wrapper function inside decorators should also be counted
+            - Tuple of a module (or filename) and qualified name (qualname).
+                Note that the qualname in the module should exist and be unique.
+                You can also use Unix shell-style wildcards to match the
+                qualname. You may also used wildcards to match the qualnames.
```

Comment:
```suggestion
            - A function. If it looks like it might be a decorated function,
                a `MaybeDecoratedFunctionWarning` will be shown.
            - Tuple of a function and a number of additional frames that should
               be skipped just before reaching this function in the stack.
               This is typically used for functions that have been decorated with
               a 'classic' decorator that replaces the function with a wrapper.
               In that case each such decorator involved should be counted in
               the number that's the second element of the tuple.
            - Tuple of a module (or filename) and qualified name (qualname).
                You can use Unix shell-style wildcards to match the qualname.
                Otherwise the qualname must appear exactly once in the
                module/file.
```